### PR TITLE
Try to fix training Loss inconsistent after resume from old checkpoint

### LINKED ISSUE #25340: Training Loss inconsistent after resume from old checkpoint
### System Info

- `transformers` version: 4.31.0
- Platform: Linux-5.4.119-19.0009.28-x86_64-with-glibc2.35
- Python version: 3.10.6
- Huggingface_hub version: 0.15.1
- Safetensors version: 0.3.1
- Accelerate version: 0.21.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.0.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

- `Accelerate` version: 0.21.0
- Platform: Linux-5.4.119-19.0009.28-x86_64-with-glibc2.35
- Python version: 3.10.6
- Numpy version: 1.22.2
- PyTorch version (GPU?): 2.0.0 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 1877.62 GB
- GPU type: NVIDIA H800
- `Accelerate` default config:
	Not found

--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.0
 [WARNING]  using untested triton version (2.0.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/usr/local/lib/python3.10/dist-packages/torch']
torch version .................... 2.0.0
deepspeed install path ........... ['/usr/local/lib/python3.10/dist-packages/deepspeed']
deepspeed info ................... 0.9.5, unknown, unknown
torch cuda version ............... 12.1
torch hip version ................ None
nvcc version ..................... 12.1
deepspeed wheel compiled w. ...... torch 2.0, cuda 12.1

### Who can help?

_No response_

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. run [run_clm.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_clm.py)  for a while. seed =42, dataset_seed = 42. model llma-7b-hf
2. start training with middle checkpoint. 
3. see the training loss. 
4. my training loss look like following:
5. Screen Shot 2023-08-07 at 2 06 58 PM
6. u can see that first resume loss is ok, but for the second the loss is inconsistent 

### Expected behavior

Training loss should be the same level before and after resume.

### PR DESCRIPTION
**Pull request by dumpmemory**
# What does this PR do?

Fixes #25340 (issue)

From my side, it might relate to the RandomSampler. i just recopy the logic from 4.29.2

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @muellerzr

**Comment by muellerzr** [comment]
Hi @dumpmemory thanks! Can you do `pip install -e -U .[quality]` and run `make style; make quality` again? This should fix that failing test

**Review by muellerzr — APPROVED** [review]
Thanks! Looks good to me and my tests all pass locally

**Comment by muellerzr** [comment]
@amyeroberts feel free to merge if it looks good with you

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1782,8 +1782,33 @@ def _inner_training_loop(
...
+                        break
+                else:
+                    # Otherwise we need to call the whooooole sampler cause there is some random operation added
+                    # AT THE VERY END!
+                    sampler = []
+
+                    if hasattr(train_dataloader, "sampler") and isinstance(train_dataloader.sampler, RandomSampler):
+                        sampler = train_dataloader.sampler
+                    else:
+                        sampler = train_dataloader.batch_sampler
```
Comment (amyeroberts) [inline src/transformers/trainer.py:1809]: This is missing the above logic where the sampler can also be `train_dataloader.batch_sample.batch_sampler.sampler`.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1782,8 +1782,33 @@ def _inner_training_loop(
...
+                        hasattr(train_dataloader, "batch_sampler")
+                        and isinstance(train_dataloader.batch_sampler.sampler, RandomSampler)
+                    )
+                    or (
+                        hasattr(train_dataloader, "batch_sampler")
+                        and hasattr(train_dataloader.batch_sampler, "batch_sampler")
+                        and isinstance(train_dataloader.batch_sampler.batch_sampler.sampler, RandomSampler)
+                    )
+                )
+                if not is_random_sampler:
```
Comment (amyeroberts) [inline src/transformers/trainer.py:1797]: This isn't quite the same as the previous logic, as there isn't a check for `is_torch_less_than_1_11`. We still support torch >= 1.9, so this should be added back too

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on fixing this! Overall change looks OK, however the logic should be simplified. 

@muellerzr what was the reason for removing this logic originally?

**Comment by muellerzr** [comment]
Originally we had thought Accelerate handled this, but it turns out it does not

**Comment by muellerzr** [comment]
@dumpmemory please have a bit of patience, our team works across multiple timezones and have many other PR's and responsibilities to get to aside this one. We'll get to this when we can, please don't spam :) Thanks

File: src/transformers/trainer_pt_utils.py [hunk 3]
```diff
@@ -55,6 +55,14 @@
 logger = logging.get_logger(__name__)
 
 
+def check_dataloader_randomsampler(dataloader):
+    if hasattr(dataloader, "sampler") and isinstance(dataloader.sampler, RandomSampler):
+        return dataloader.sampler, True
+    if hasattr(dataloader, "batch_sampler"):
+        return check_dataloader_randomsampler(dataloader.batch_sampler)
+    return dataloader.sampler, False
```
Comment (amyeroberts) [inline src/transformers/trainer_pt_utils.py:63]: This should just return the sampler, and then the user can choose what they do with the output e.g. check if it's a random sampler. This ensures the function is as versatile as possible and can be used / extended without issue. 


```suggestion
def get_dataloader_sampler(dataloader):
    if hasattr(dataloader, "sampler"):
        return dataloader.sampler
    if hasattr(dataloader, "batch_sampler"):
        return get_dataloader_sampler(dataloader.batch_sampler)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating - code it looking good! 

Just a comment on the utility function - we want functions to be as atomic as possible. Once updated we'll be good to merge.

File: src/transformers/trainer_pt_utils.py [hunk 4]
```diff
@@ -55,6 +55,15 @@
 logger = logging.get_logger(__name__)
 
 
+def get_dataloader_sampler(dataloader):
+    # after accelerate.prepare function the wraped dataloader.sampler will be SequentialSampler instead of RandomSampler
+    if hasattr(dataloader, "sampler") and isinstance(dataloader.sampler, RandomSampler):
+        return dataloader.sampler
+    if hasattr(dataloader, "batch_sampler") and dataloader.batch_sampler is not None:
+        return get_dataloader_sampler(dataloader.batch_sampler)
+    return dataloader.sampler
```
Comment (amyeroberts) [inline src/transformers/trainer_pt_utils.py:64]: * In the last line, you can't return `dataloader.sampler` if there dataloader doesn't have that attribute. 
* The `isinstance` check isn't needed. You just want to return any sampler and then the user can test if they want. In fact, the current logic will already do that: it would return a random sample because it would evaluate as false for the if checks and then finally return `dataloader.sampler`

```suggestion
    if hasattr(dataloader, "sampler"):
        return dataloader.sampler
    if hasattr(dataloader, "batch_sampler") and dataloader.batch_sampler is not None:
        return get_dataloader_sampler(dataloader.batch_sampler)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating on this. Just one last comment on the structure of `get_dataloader_sampler`

File: src/transformers/trainer_pt_utils.py [hunk 5]
```diff
@@ -55,6 +55,17 @@
...
 
+def get_dataloader_sampler(dataloader):
+    # after accelerate.prepare function the wraped dataloader.sampler will be SequentialSampler instead of RandomSampler
+    if hasattr(dataloader, "sampler") and isinstance(dataloader.sampler, RandomSampler):
+        return dataloader.sampler
+    elif hasattr(dataloader, "batch_sampler") and dataloader.batch_sampler is not None:
+        return get_dataloader_sampler(dataloader.batch_sampler)
+    elif hasattr(dataloader, "sampler"):
+        return dataloader.sampler
+    return None
```
Comment (amyeroberts) [inline src/transformers/trainer_pt_utils.py:66]: @dumpmemory Could you please explain why you keep adding the `isinstance` check into this function? In the third logic check `elif  hasattr(dataloader, "sampler"):` means we will still be returning SequentialSampler. As I mentioned in the previous review, it's not needed. Neither is the final `return None` - this already happens in the function. 

If it's the case that when SequentialSampler is set, then the dataloader can also have a batch_sampler that is None and a sampler, we can just change the order of checks.

```suggestion
def get_dataloader_sampler(dataloader):
    if hasattr(dataloader, "batch_sampler") and dataloader.batch_sampler is not None:
        return get_dataloader_sampler(dataloader.batch_sampler)
    elif hasattr(dataloader, "sampler"):
        return dataloader.sampler
```

**Review by amyeroberts — COMMENTED** [review]
@dumpmemory Could you explain in some more detail why the suggested implementation of `get_dataloader_sampler` isn't the one being used? For the current diff, it's not clear why some of the additional logic e.g. checking isinstance is added.

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating!

**Review by muellerzr — APPROVED** [review]
Great work!
